### DISCUSSION TITLE
Issues with importing own data into NNSM

### DISCUSSION THREAD

**Discussion by epergant** [pnnl/neuromancer#138] [Q&A] [answered]
Hello, I have a test dataset for a building on the Purdue campus which I want to integrate into Neuromancer to develop a DPC. This location has already had an MPC demonstration, so it would be nice to compare in simulation th
[Test Neuromancer.zip](https://github.com/pnnl/neuromancer/files/14625491/Test.Neuromancer.zip)
e performance of DPC. 

I have attached my dataset:
[Test Neuromancer.zip](https://github.com/pnnl/neuromancer/files/14625492/Test.Neuromancer.zip)
[DC_House_Nov_April.xlsx](https://github.com/pnnl/neuromancer/files/14625488/DC_House_Nov_April.xlsx)

I need some help with how to integrate my file with the PSL format are there any tutorials on this? I have attempted rewriting the code in: https://github.com/pnnl/neuromancer/blob/master/examples/domain_examples/NSSM_building_dynamics.py, to integrate my data, see attached Python code but with little training success. Currently my dynamic model is of the form:

x(k+1) = A*x(k) + B*u(k) + D*d(k)
y(k) = C * x(k)

where C = 1, i.e. we only measure and want to control the mixed air temperature of the house, and we do not incorporate any noise terms. u = [Heat Supply; Outdoor Temperature], and d = [Solar Gains; Wind Speed]. I have previously trained gray-box models with this dataset, it would be nice to extend to a black box.

Any help is appreciated.
Elias

**Comment by drgona** [comment, marked as answer]
Hi @epergant, thanks for your interest in our library.

In general, you can load your csv file as Python dictionary with arbitrary string values for keys and numpy arrays as values.
See [these](https://blog.finxter.com/5-best-ways-to-convert-csv-to-dictionary-in-python/) general ways to load csv files into Python dict. 

After you have your data in Python dictionary, you don't need to be compliant with our specific PSL format, where we use specific keys. You can use any keys you want and build your dataset around those.
If you want to use the PSL format, you need to categorize your variables into one of these keys: 
- 'x' as for full state space
- 'y' for controlled states
- 'u' for control actions
- 'u' for disturbances
- 'Time' for time sequence
- 'exp_id' for index of the experiment (batch)
If you have multiple variables, you need to index them such as 'x_1', 'x_2', etc. But again you are free to use your own labels, e.g. 'Heat Supply', 'Outdoor Temperature' instead of 'u1', and 'u2'.

If you choose to use defauls PSL labels, then you could use [PSL file_emulator](https://github.com/pnnl/neuromancer/blob/master/src/neuromancer/psl/file_emulator.py) to load your formatted csv file. Specifically you could use [_read_file](https://github.com/pnnl/neuromancer/blob/master/src/neuromancer/psl/file_emulator.py#L46) function that will parse this PSL format dataset from csv or mat file.

Regarding the model architecture, the nice thing about Neuromancer is that you are not limited to linear state space models. You can use wide-range of black- or gray- box models from our library or develop your own. See some examples here:
https://github.com/pnnl/neuromancer/tree/master?tab=readme-ov-file#ordinary-differential-equations-odes

Let me know if this helps.
